### PR TITLE
Refactor distributed GPU device initialization.

### PR DESCRIPTION
**Pull request by hawkinsp**
Avoid reregistering backend factories; instead simply have the usual
factory function support distributed GPU.

### CODE REVIEW

File: jax/_src/distributed.py [hunk 1]
```diff
@@ -26,6 +26,7 @@
 from jax._src.lib import xla_extension
 
 class State:
+  process_id: int = 0
```
Comment (yashk2810) [inline jax/_src/distributed.py:29]: Shouldn't this be None as default to match the current behavior?

Comment (hawkinsp) [reply jax/_src/distributed.py:29]: It doesn't greatly matter. If you pass `None` to the underlying client init function, it defaults to 0. This is simple: always an int (after all, you have to have a node id, even if not distributed.)

File: jax/_src/distributed.py [hunk 2]
```diff
@@ -34,8 +35,9 @@ def initialize(self,
                  coordinator_address: Optional[str] = None,
                  num_processes: Optional[int] = None,
                  process_id: Optional[int] = None):
-    coordinator_address = os.environ.get('JAX_COORDINATOR_ADDRESS',
-                                         None) or coordinator_address
+    self.process_id = process_id
+    coordinator_address = (coordinator_address or
```
Comment (hawkinsp) [inline jax/_src/distributed.py:39]: Note this change. This is intentional: if someone specifies a specific value for an arg, an environment variable should not override it.

File: jax/_src/lib/xla_bridge.py [hunk 3]
```diff
@@ -210,17 +211,23 @@ def register_backend_factory(name, factory, *, priority=0):
 register_backend_factory('tpu_driver', _make_tpu_driver_client,
                          priority=100)
 
+
+def _make_gpu_client(platform_name=None):
+  return xla_client.make_gpu_client(
```
Comment (yashk2810) [inline jax/_src/lib/xla_bridge.py:216]: You should default to 0 here:

if distributed.global_state.process_id is None:
  node_id = 0
else:
  node_id = distributed.global_state.process_id

File: jax/_src/distributed.py (hunk 1 above)
Comment (yashk2810) [reply jax/_src/distributed.py:29]: The current behavior is to error if process_id is None.

For TPUs we figure that thing out.

File: jax/_src/lib/xla_bridge.py (hunk 3 above)
Comment (hawkinsp) [reply jax/_src/lib/xla_bridge.py:216]: Isn't it just fewer lines of code to default to `0` in `distributed.py`?

Comment (yashk2810) [reply jax/_src/lib/xla_bridge.py:216]: Right.. but there are 2 behaviors right now and I am confused about them.

initialize has the default None which raises an error if you don't specify process_id.

But the class variable defaults to 0. Previously, process_id was a required field for GPUs.

Comment (hawkinsp) [reply jax/_src/lib/xla_bridge.py:216]: There are two cases: single-process and distributed:
* In the single-process case, you don't call the distributed init function. The default is 0, which is the correct default.
* In the distributed case, call the init function. You must specify a `process_id` (the init function enforces this), unless you are on cloud TPU in which case it is inferred.

**Review by yashk2810 — APPROVED** [review]

**Pull request merged**